File: A2/Task1.py
import pandas as pd
from datetime import datetime


# Task 1-2
class ClimateData:
    def __init__(
        self,
        local_date,
        speed_max_gust,
        total_precipitation,
        min_temp,
        max_temp,
        mean_temp,
    ):
        self.local_date = local_date
        self.speed_max_gust = speed_max_gust
        self.total_precipitation = total_precipitation
        self.min_temp = min_temp
        self.max_temp = max_temp
        self.mean_temp = mean_temp

    def __str__(self):
        return (
            f"Date: {self.local_date}, "
            f"Max Gust: {self.speed_max_gust}, "
            f"Precipitation: {self.total_precipitation}, "
            f"Min Temp: {self.min_temp}, "
            f"Max Temp: {self.max_temp}, "
            f"Mean Temp: {self.mean_temp}"
        )


# Convert Data into correct format to prase
def parse_date(date_str):
    try:
        return datetime.strptime(date_str, "%Y-%m-%d %H:%M").date()
    except ValueError:
        return None


# Convert value to float to parse
def parse_numeric(value, default=None):
    try:
        return float(value)
    except ValueError:
        return default


def read_and_process_csv(file_path):
    df = pd.read_csv(file_path)
    climate_data_list = []

    for index, row in df.drop_duplicates().iterrows():
        # If no data in column then skip it
        if pd.notna(row["SPEED_MAX_GUST"]):
            # Parse data and added them to the list
            climate_data = ClimateData(
                local_date=parse_date(row["LOCAL_DATE"]),
                speed_max_gust=parse_numeric(row["SPEED_MAX_GUST"]),
                total_precipitation=parse_numeric(row["TOTAL_PRECIPITATION"]),
                min_temp=parse_numeric(row["MIN_TEMPERATURE"]),
                max_temp=parse_numeric(row["MAX_TEMPERATURE"]),
                mean_temp=parse_numeric(row["MEAN_TEMPERATURE"]),
            )
            climate_data_list.append(climate_data)

            # Print data //testing
            # for data in climate_data_list:
            #     print(data)
            climate_data_df = pd.DataFrame([vars(data) for data in climate_data_list])

    return climate_data_df


# Task 3
def find_max_precipitation_month(climate_data_df):
    # Convert 'local_date' to datetime
    climate_data_df["local_date"] = pd.to_datetime(climate_data_df["local_date"])

    # Group by month and year, sum the total precipitation values
    monthly_precipitation = climate_data_df.groupby(
        [climate_data_df["local_date"].dt.year, climate_data_df["local_date"].dt.month]
    )["total_precipitation"].sum()

    # Find the month-year combination with the highest total precipitation
    max_precipitation_month = monthly_precipitation.idxmax()
    max_precipitation_value = monthly_precipitation.max()
    # Extract year and month from the tuple
    max_precipitation_year, max_precipitation_month = max_precipitation_month

    return max_precipitation_year, max_precipitation_month, max_precipitation_value


def find_max_gust_day(climate_data_df):
    # Find the day with the highest SPEED_MAX_GUST
    max_gust_day = climate_data_df.loc[climate_data_df["speed_max_gust"].idxmax()]

    return max_gust_day


def find_max_fluctuation(climate_data_df):
    # Calculate temperature fluctuation for each day
    climate_data_df["temperature_fluctuation"] = (
        climate_data_df["max_temp"] - climate_data_df["min_temp"]
    )

    # Find the day with the largest temperature fluctuation and round to 1 decimal
    max_fluctuation_day = climate_data_df.loc[
        climate_data_df["temperature_fluctuation"].idxmax()
    ]
    round_max = round(max_fluctuation_day["temperature_fluctuation"], 2)

    return max_fluctuation_day, round_max


# Task 4
def generate_monthly_averages(climate_data_df):
    climate_data_df["local_date"] = pd.to_datetime(climate_data_df["local_date"])
    # Group by year and month and calculate averages
    monthly_averages = climate_data_df.groupby(
        [climate_data_df["local_date"].dt.year, climate_data_df["local_date"].dt.month]
    ).agg(
        {
            "speed_max_gust": "mean",
            "total_precipitation": "mean",
            "min_temp": "mean",
            "max_temp": "mean",
            "mean_temp": "mean",
        }
    )
    # Change format of date to YYYY-MM-DD
    monthly_averages.index = monthly_averages.index.map(
        lambda x: f"{x[0]}-{x[1]:02d}-01"
    )

    return monthly_averages


def generate_records_between_dates(climate_data_df, start_date, end_date):
    start_date = parse_date(start_date)
    end_date = parse_date(end_date)

    if start_date is None or end_date is None:
        print("Invalid date format. Please provide dates in the format 'YYYY-MM-DD'.")
        return None

    # Ensure the dates are within the range of the dataset
    if not (
        climate_data_df["local_date"].min()
        <= start_date
        <= climate_data_df["local_date"].max()
        and climate_data_df["local_date"].min()
        <= end_date
        <= climate_data_df["local_date"].max()
    ):
        print("Provided dates are outside the range of the dataset.")
        return None

    # Filter records
    filtered_records = climate_data_df[
        (climate_data_df["local_date"] >= start_date)
        & (climate_data_df["local_date"] <= end_date)
    ]

    return filtered_records


# Testing
def main():
    file_path = "/Users/orz_os/Desktop/CP472/Kham0130_a02/climate-daily.csv"
    climate_data_df = read_and_process_csv(file_path)
    max_precipitation_year, max_precipitation_month, max_precipitation_value = (
        find_max_precipitation_month(climate_data_df)
    )
    max_gust_day = find_max_gust_day(climate_data_df)
    max_fluctuation_day, round_max = find_max_fluctuation(climate_data_df)

    print(
        f"Month with the highest total precipitation: {max_precipitation_year}-{max_precipitation_month}, "
        f"Total Precipitation: {max_precipitation_value}"
    )
    print(
        f"Day with the highest SPEED_MAX_GUST: {max_gust_day['local_date'].date()}, Speed: {max_gust_day['speed_max_gust']}"
    )
    print(
        f"Day with the largest temperature fluctuation: {max_fluctuation_day['local_date'].date()}, "
        f"Temperature Fluctuation: {round_max}"
    )

    while True:
        print("\nSelect a report to generate:")
        print("1. Monthly Averages")
        print("2. Records Between Dates")
        print("3. Quit")

        choice = input(
            "Enter the number of the report you want to generate (or 'q' to quit): "
        )

        if choice.lower() == "q":
            print("Exiting the program.")
            break
        elif choice == "1":
            monthly_averages = generate_monthly_averages(climate_data_df)
            print("\nMonthly Averages:")
            print(monthly_averages)
        elif choice == "2":
            # When entering input, it needs to be in the exact format as shown in brackets with time default to 0:00
            start_date = input("Enter the start date (YYYY-MM-DD 0:00): ")
            end_date = input("Enter the end date (YYYY-MM-DD 0:00): ")
            records_between_dates = generate_records_between_dates(
                climate_data_df, start_date, end_date
            )
            if records_between_dates is not None:
                print("\nRecords Between Dates:")
                print(records_between_dates)
            else:
                print("Invalid choice. Please enter 1, 2, or 'q'.")


if __name__ == "__main__":
    main()

File: A2/Task1.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define MAX_LINE_LENGTH 1024

// Task 1-2
struct ClimateRecord
{
    char localDate[11];
    double speedMaxGust;
    double totalPrecipitation;
    double minTemperature;
    double maxTemperature;
    double meanTemperature;
};

int parseCSVLine(char *line, struct ClimateRecord *record)
{
    char *token = strtok(line, ",");
    if (token == NULL)
        return 0;

    // Extract only the date part
    strncpy(record->localDate, token, 10);
    record->localDate[10] = '\0';

    token = strtok(NULL, ",");
    if (token == NULL)
        return 0;
    record->speedMaxGust = atof(token);

    token = strtok(NULL, ",");
    if (token == NULL)
        return 0;
    record->totalPrecipitation = atof(token);

    token = strtok(NULL, ",");
    if (token == NULL)
        return 0;
    record->minTemperature = atof(token);

    token = strtok(NULL, ",");
    if (token == NULL)
        return 0;
    record->maxTemperature = atof(token);

    token = strtok(NULL, ",");
    if (token == NULL)
        return 0;
    record->meanTemperature = atof(token);

    return 1;
}

// Task 3
void findMaxPrecipitationMonth(struct ClimateRecord *records, size_t size)
{
    double maxPrecipitation = 0;
    char maxPrecipitationMonth[8];

    int currentYear, currentMonth;
    double currentMonthTotal = 0;

    for (size_t i = 0; i < size; ++i)
    {
        // Extract year and month from the date (assuming date format is YYYY-MM-DD)
        int year, month;
        sscanf(records[i].localDate, "%d-%d", &year, &month);

        // Check if it's a new month
        if (i == 0 || year != currentYear || month != currentMonth)
        {
            if (currentMonthTotal > maxPrecipitation)
            {
                maxPrecipitation = currentMonthTotal;
                sprintf(maxPrecipitationMonth, "%02d-%04d", currentMonth, currentYear);
            }

            // Reset total for the new month
            currentYear = year;
            currentMonth = month;
            currentMonthTotal = records[i].totalPrecipitation;
        }
        else
        {
            // Accumulate total for the current month
            currentMonthTotal += records[i].totalPrecipitation;
        }
    }

    // Check the last month
    if (currentMonthTotal > maxPrecipitation)
    {
        maxPrecipitation = currentMonthTotal;
        sprintf(maxPrecipitationMonth, "%02d-%04d", currentMonth, currentYear);
    }

    printf("Month with the highest TOTAL_PRECIPITATION: %s, Precipitation: %.2f\n", maxPrecipitationMonth, maxPrecipitation);
}

void findMaxGust(struct ClimateRecord *records, size_t size)
{
    double maxGust = 0;
    char maxGustDay[11];

    for (size_t i = 0; i < size; ++i)
    {
        double gust = records[i].speedMaxGust;
        if (gust > maxGust)
        {
            maxGust = gust;
            strcpy(maxGustDay, records[i].localDate);
        }
    }

    printf("Day with the highest SPEED_MAX_GUST: %s, Speed: %.2f\n", maxGustDay, maxGust);
}

void findMaxTemperatureFluctuation(struct ClimateRecord *records, size_t size)
{
    double maxFluctuation = 0;
    char maxFluctuationDay[11]; // Assuming date format is YYYY-MM-DD

    for (size_t i = 0; i < size; ++i)
    {
        double fluctuation = records[i].maxTemperature - records[i].minTemperature;
        if (fluctuation > maxFluctuation)
        {
            maxFluctuation = fluctuation;
            strcpy(maxFluctuationDay, records[i].localDate);
        }
    }

    printf("Day with the largest temperature fluctuation: %s, Fluctuation: %.2f\n", maxFluctuationDay, maxFluctuation);
}

// Task 4
void generateMonthlyAverages(struct ClimateRecord *records, size_t size)
{
    printf("\nMonthly Averages:\n");

    int currentYear = -1;
    int currentMonth = -1;
    int count = 0;
    double totalSpeedMaxGust = 0;
    double totalTotalPrecipitation = 0;
    double totalMinTemperature = 0;
    double totalMaxTemperature = 0;
    double totalMeanTemperature = 0;

    for (size_t i = 0; i < size; ++i)
    {
        int year, month;
        sscanf(records[i].localDate, "%d-%d", &year, &month);

        if (currentYear == -1 || currentMonth == -1 || (currentYear == year && currentMonth == month))
        {
            // Same month or first iteration
            totalSpeedMaxGust += records[i].speedMaxGust;
            totalTotalPrecipitation += records[i].totalPrecipitation;
            totalMinTemperature += records[i].minTemperature;
            totalMaxTemperature += records[i].maxTemperature;
            totalMeanTemperature += records[i].meanTemperature;
            count++;

            // Update current month and year
            currentYear = year;
            currentMonth = month;
        }
        else
        {
            // Print averages for the previous month
            printf("Month-Year: %02d-%04d, Avg SpeedMaxGust: %.2f, Avg TotalPrecipitation: %.2f, Avg MinTemperature: %.2f, Avg MaxTemperature: %.2f, Avg MeanTemperature: %.2f\n",
                   currentMonth, currentYear,
                   totalSpeedMaxGust / count, totalTotalPrecipitation / count,
                   totalMinTemperature / count, totalMaxTemperature / count,
                   totalMeanTemperature / count);

            // Reset totals and count for the new month
            totalSpeedMaxGust = records[i].speedMaxGust;
            totalTotalPrecipitation = records[i].totalPrecipitation;
            totalMinTemperature = records[i].minTemperature;
            totalMaxTemperature = records[i].maxTemperature;
            totalMeanTemperature = records[i].meanTemperature;
            count = 1;

            // Update current month and year
            currentYear = year;
            currentMonth = month;
        }
    }

    // Print averages for the last month
    if (count > 0)
    {
        printf("Month-Year: %02d-%04d, Avg SpeedMaxGust: %.2f, Avg TotalPrecipitation: %.2f, Avg MinTemperature: %.2f, Avg MaxTemperature: %.2f, Avg MeanTemperature: %.2f\n",
               currentMonth, currentYear,
               totalSpeedMaxGust / count, totalTotalPrecipitation / count,
               totalMinTemperature / count, totalMaxTemperature / count,
               totalMeanTemperature / count);
    }
}

void filterRecordsByDate(struct ClimateRecord *records, size_t size)
{
    printf("\nEnter Start Date (YYYY-MM-DD): ");
    char startDate[11];
    scanf("%s", startDate);

    printf("Enter End Date (YYYY-MM-DD): ");
    char endDate[11];
    scanf("%s", endDate);

    printf("\nWeather Records between %s and %s:\n", startDate, endDate);

    for (size_t i = 0; i < size; ++i)
    {
        if (strcmp(records[i].localDate, startDate) >= 0 && strcmp(records[i].localDate, endDate) <= 0)
        {
            printf("Date: %s, SpeedMaxGust: %.2f, TotalPrecipitation: %.2f, MinTemperature: %.2f, MaxTemperature: %.2f, MeanTemperature: %.2f\n",
                   records[i].localDate, records[i].speedMaxGust, records[i].totalPrecipitation,
                   records[i].minTemperature, records[i].maxTemperature, records[i].meanTemperature);
        }
    }
}

// Testing
int main()
{
    FILE *file = fopen("/Users/orz_os/Desktop/CP472/Kham0130_a02/climate-daily.csv", "r");
    if (file == NULL)
    {
        perror("Error opening file");
        return 1;
    }

    // Create a dynamic array to store ClimateRecord structs
    struct ClimateRecord *records = NULL;
    size_t capacity = 0;
    size_t size = 0;

    char line[MAX_LINE_LENGTH];
    while (fgets(line, sizeof(line), file))
    {
        // Remove newline character if present
        size_t len = strlen(line);
        if (len > 0 && line[len - 1] == '\n')
        {
            line[len - 1] = '\0';
        }

        // Allocate memory for a new ClimateRecord
        struct ClimateRecord *newRecord = malloc(sizeof(struct ClimateRecord));
        if (newRecord == NULL)
        {
            perror("Error allocating memory");
            return 1;
        }

        // Parse the CSV line and populate the ClimateRecord
        if (parseCSVLine(line, newRecord))
        {
            // Add the new record to the dynamic array
            if (size == capacity)
            {
                capacity = (capacity == 0) ? 1 : capacity * 2;
                records = realloc(records, capacity * sizeof(struct ClimateRecord));
                if (records == NULL)
                {
                    perror("Error reallocating memory");
                    return 1;
                }
            }

            records[size++] = *newRecord;
        }
        else
        {
            // Skip malformed or missing data
            free(newRecord);
        }
    }

    // Close the file
    fclose(file);

    findMaxPrecipitationMonth(records, size);
    findMaxGust(records, size);
    findMaxTemperatureFluctuation(records, size);

    int choice;
    do
    {
        printf("\nChoose a Report to Generate:\n");
        printf("1. Monthly Averages\n");
        printf("2. Weather Records between a Given Start and End Date\n");
        printf("0. Exit\n");
        printf("Enter your choice (0-2): ");
        scanf("%d", &choice);

        switch (choice)
        {
        case 1:
            generateMonthlyAverages(records, size);
            break;
        case 2:
            filterRecordsByDate(records, size);
            break;
        case 0:
            printf("Exiting...\n");
            break;
        default:
            printf("Invalid choice. Please enter a number between 0 and 2.\n");
        }
    } while (choice != 0);

    // Free allocated memory
    free(records);

    return 0;
}
